corgi/tasks/prod_defs.py

```
@@ -342,8 +422,8 @@ def parse_errata_info(
     """Parse and create ProductVariants from errata_info in product-definitions.json"""
     et_product_versions_set = set(product_stream.et_product_versions)
     for et_product in errata_info:
-        et_product_name = et_product.pop("product_name")
-        et_product_versions = et_product.pop("product_versions")
+        et_product_name = et_product.get("product_name", "")
+        et_product_versions = et_product.get("product_versions", [])
```

The old code makes sure that we stop processing with an error if the data in prod-defs is formatted incorrectly. The new code you have here will silently ignore any missing product_name or product_versions keys in some stream's errata_info. That seems surprising to me...is there a reason we need this change?

---

The reason is that we reuse the dict again at the end of update_product function to reconcile what's in the Corgi DB with what we just read from product_definitions. If we use pop then all the keys are missing from the original dictionary.

---

Maybe we should leave this change for a separate MR. But I think it's better to keep the .pop() we had in the old code, and fail in Corgi / this task if some value is missing in prod-defs.

We could create a new dict (instead of reusing this one for multiple purposes) to keep track of what we read from prod-defs, and use that at the end to check for orphaned objects. But I also don't think a dict is needed.

We do update_or_create for each ProductModel. After the update_or_create succeeds, we can add the created / updated ProductModel's primary key to a set for that ProductModel type.

At the end of the update_products() task, we can do a query to find all the existing ProductModel objects of a certain type, which are not in the set / were not updated or created. That means they don't exist in prod-defs anymore, so we can just .delete() that queryset. For example:

```
# At the very start of the task
prod_defs_products = set()
prod_defs_product_versions = set()
# and so on
...
# When creating e.g. a ProductVersion
product_version, created = ProductVersion.objects.update_or_create(
    # omitted for clarity
)
# We add the PK to the set, because we processed it
# So this version must still be present in prod-defs data
prod_defs_product_versions.add(product_version.pk)
...
# At the very end of the task
# Any ProductVersions in Corgi, which were not created / updated / processed by this task at all,
# must no longer be defined in prod-defs, so we can delete them in Corgi
orphanded_corgi_product_versions = ProductVersion.objects.exclude(pk__in=prod_defs_product_versions)
orphaned_corgi_product_versions.delete()
```

---

Ok, I'll remove this clean up code for now, and create a separate issue to follow up with this after.

---

Created CORGI-784